include/rmm/mr/device/arena_memory_resource.hpp

```diff
@@ -0,0 +1,287 @@
+ * allocation: A survey and critical review. In International Workshop on Memory Management (pp.
+ * 1-116). Springer, Berlin, Heidelberg.
+ * \see Berger, E. D., McKinley, K. S., Blumofe, R. D., & Wilson, P. R. (2000). Hoard: A scalable
+ * memory allocator for multithreaded applications. ACM Sigplan Notices, 35(11), 117-128.
+ * \see Evans, J. (2006, April). A scalable concurrent malloc (3) implementation for FreeBSD. In
+ * Proc. of the bsdcan conference, ottawa, canada.
+ * \see https://sourceware.org/glibc/wiki/MallocInternals
+ * \see http://hoard.org/
+ * \see http://jemalloc.net/
+ * \see https://github.com/google/tcmalloc
+ *
+ * @tparam UpstreamResource memory_resource to use for allocating the arenas. Implements
+ * rmm::mr::device_memory_resource interface.
+ */
+template <typename Upstream>
+class arena_memory_resource final : public device_memory_resource {
+ public:
+  /// The required alignment of this allocator.
+  static constexpr std::size_t allocation_alignment = 256;
+
+  /**
+   * @brief Construct an `arena_memory_resource`.
+   *
+   * @throws rmm::logic_error if `upstream_mr == nullptr`
+   *
+   * @param upstream_mr The memory_resource from which to allocate memory for the arenas.
+   */
+  explicit arena_memory_resource(Upstream* upstream_mr) : upstream_mr_{upstream_mr}
+  {
+    RMM_EXPECTS(nullptr != upstream_mr, "Unexpected null upstream pointer.");
+  }
+
+  // Disable copy (and move) semantics.
+  arena_memory_resource(arena_memory_resource const&) = delete;
+  arena_memory_resource& operator=(arena_memory_resource const&) = delete;
+
+  /**
+   * @brief Queries whether the resource supports use of non-null CUDA streams for
+   * allocation/deallocation.
+   *
+   * @returns bool true.
+   */
+  bool supports_streams() const noexcept override { return true; }
+
+  /**
+   * @brief Query whether the resource supports the get_mem_info API.
+   *
+   * @return bool false
+   */
+  bool supports_get_mem_info() const noexcept override { return false; }
+
+ private:
+  using arena      = detail::arena::arena<Upstream>;
+  using read_lock  = std::shared_lock<std::shared_timed_mutex>;
+  using write_lock = std::lock_guard<std::shared_timed_mutex>;
+
+  /**
+   * @brief Allocates memory of size at least `bytes`.
+   *
+   * The returned pointer has at least 256B alignment.
+   *
+   * @throws `std::bad_alloc` if the requested allocation could not be fulfilled
+   *
+   * @param bytes The size in bytes of the allocation
+   * @param stream The stream to associate this allocation with
+   * @return void* Pointer to the newly allocated memory
+   */
+  void* do_allocate(std::size_t bytes, cudaStream_t stream) override
+  {
+    if (bytes <= 0) return nullptr;
+
+    bytes = rmm::detail::align_up(bytes, allocation_alignment);
+    RMM_EXPECTS(
+      bytes <= get_maximum_allocation_size(), rmm::bad_alloc, "Maximum allocation size exceeded");
```

Comment:
Since `get_maximum_allocation_size` returns the largest possible `size_t`, it's impossible for this expectation to fail. I don't think this is needed.

Reply:
Removed.